### PR TITLE
Product exporter plugin

### PR DESCRIPTION
**Pull request by xphere**
Simple product export plugin.
Listens to `bamboo_admin.table_buttons` event and renders a button to export products in csv format.

### CODE REVIEW

File: src/Elcodi/Plugin/ProductCsv/composer.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+            "name": "Berny Cantos",
+            "email": "be@rny.cc"
+        },
+        {
+            "name": "Elcodi Community",
+            "homepage": "https://github.com/elcodi-plugin/product-csv/contributors"
+        }
+    ],
+    "require": {
+        "goodby/csv": "~1.0",
```
Comment (mmoreram) [inline src/Elcodi/Plugin/ProductCsv/composer.json]: php5.4 & other dependencies (elcodi/core-bundle..., elcodi/product-bundle)

File: src/Elcodi/Plugin/ProductCsv/Resources/config/twig.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+services:
+
+    #
+    # Twig renderer
+    #
+    elcodi.plugin.product_csv.renderer:
+        class: Elcodi\Plugin\ProductCsv\Templating\TwigRenderer
+        arguments:
+            - @twig
+            - @elcodi.hook_system
```
Comment (mmoreram) [inline src/Elcodi/Plugin/ProductCsv/Resources/config/twig.yml]: This service should not be injected :dancer:

File: src/Elcodi/Plugin/ProductCsv/composer.json (hunk 1 above)
Comment (xphere) [inline src/Elcodi/Plugin/ProductCsv/composer.json]: Those are direct dependencies for `elcodi/plugin-bundle`, not for this plugin. Should we require dependencies of dependencies?

File: src/Elcodi/Plugin/ProductCsv/Resources/config/twig.yml (hunk 2 above)
Comment (xphere) [inline src/Elcodi/Plugin/ProductCsv/Resources/config/twig.yml]: Fixed! Thx

File: src/Elcodi/Plugin/ProductCsv/Controller/CsvController.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        $exporter = $this->get('elcodi.plugin.product_csv.exporter');
+
+        $response = new StreamedResponse();
+        $response->setCallback(function () use ($repository, $exporter) {
+            $rows = $repository->findAll();
+
+            $exporter->export($rows);
+        });
+
+        return $this->makeDownloadable($response, 'products.csv', 'text/csv');
```
Comment (mmoreram) [inline src/Elcodi/Plugin/ProductCsv/Controller/CsvController.php]: Maybe the file name should be defined in config?

File: src/Elcodi/Plugin/ProductCsv/composer.json (hunk 1 above)
Comment (mmoreram) [inline src/Elcodi/Plugin/ProductCsv/composer.json]: 5.4 is always required by all repositories.

And then, do we have some Symfony components namespaces in this repository

```php
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Method;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
use Symfony\Bundle\FrameworkBundle\Controller\Controller;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpFoundation\ResponseHeaderBag;
use Symfony\Component\HttpFoundation\StreamedResponse;
```

And services requirements:

``` php
$repository = $this->get('elcodi.repository.product');
```

So this means that all this dependencies must be defined in a hard way in the plugin.

Comment (xphere) [inline src/Elcodi/Plugin/ProductCsv/composer.json]: Right, my bad. I'm looking for classes used in files changed.
```shell
grep -hw ^use --include=*.php -- $(git diff master..HEAD --name-only) | sort
use DateTime;
use Elcodi\Bundle\CoreBundle\DependencyInjection\Abstracts\AbstractExtension;
use Elcodi\Component\Plugin\EventInterface;
use Elcodi\Component\Product\Entity\Interfaces\ProductInterface;
use Elcodi\Plugin\ProductCsv\DependencyInjection\ProductCsvExtension;
use Goodby\CSV\Export\Protocol\ExporterInterface;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Method;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
use Symfony\Bundle\FrameworkBundle\Controller\Controller;
use Symfony\Component\DependencyInjection\Extension\ExtensionInterface;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpFoundation\ResponseHeaderBag;
use Symfony\Component\HttpFoundation\StreamedResponse;
use Symfony\Component\HttpKernel\Bundle\Bundle;
use Twig_Environment;
```
So we need to require, at least, `elcodi/product`, `symfony/framework-extra-bundle`, `symfony/framework-bundle`, `symfony/http-foundation`, `symfony/http-kernel` and `twig/twig`.

As it needs the `elcodi.repository.product` service, replace `elcodi/product` with `elcodi/product-bundle`.

And require `symfony/dependency-injection` as @mmoreram suggests, so the default adapter works flawlessly.

**Comment by xphere** [comment]
@hacfi yes, this were meant as a proof of concept for the new plugin feature, so it is rather simplified. The default setup works with utf file input and output. Maybe the next step is to allow settings through the new configuration-bundle.

**Comment by mmoreram** [comment]
@xphere Please, rebase this feature and do something with it! :)

**Comment by xphere** [comment]
We need to talk more about the pluginization of our system before accepting this. I'll set the PR on hold.

**Comment by mmoreram** [comment]
@xphere ASAP this PR is ready with new plugin changes, please ping me :)

**Comment by xphere** [comment]
Rebased and working as expected now with the new Plugin system.

**Comment by xphere** [comment]
:dancer:

**Pull request merged**